src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,931 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "corpriv.h"
+#include "tailcallhelp.h"
+#include "dllimport.h"
+#include "formattype.h"
+#include "sigformat.h"
+#include "gcrefmap.h"
+#include "threads.h"
+
+#ifndef CROSSGEN_COMPILE
+
+FCIMPL2(void*, TailCallHelp::AllocTailCallArgBuffer, INT32 size, void* gcDesc)
+{
+    FCALL_CONTRACT;
+
+    _ASSERTE(size >= 0);
+
+    return GetThread()->AllocTailCallArgBuffer(static_cast<size_t>(size), gcDesc);
+}
+FCIMPLEND
+
+FCIMPL0(void, TailCallHelp::FreeTailCallArgBuffer)
+{
+    FCALL_CONTRACT;
+    GetThread()->FreeTailCallArgBuffer();
+}
+FCIMPLEND
+
+FCIMPL2(void*, TailCallHelp::GetTailCallInfo, void** retAddrSlot, void** retAddr)
+{
+    FCALL_CONTRACT;
+
+    Thread* thread = GetThread();
+
+    *retAddr = thread->GetReturnAddress(retAddrSlot);
+    return thread->GetTailCallTls();
+}
+FCIMPLEND
+
+#endif
+
+struct ArgBufferValue
+{
+    TypeHandle TyHnd;
+    unsigned int Offset;
+
+    ArgBufferValue(TypeHandle tyHnd = TypeHandle(), unsigned int offset = 0)
+        : TyHnd(tyHnd), Offset(offset)
+    {
+    }
+};
+
+struct ArgBufferLayout
+{
+    bool HasTargetAddress;
```

Comment:
What is the benefit of having HasTargetAddress and TargetAddressOffset instead of just a TargetAddress field with NULL indicating it doesn't have target address?

Reply:
This is just the offset into the arg buffer so it is not really a pointer. `0` is a valid possible value. I could change it to have `0xFFFFFFFF` represent an invalid value instead -- would you think that's better?

Reply:
I meant not storing the address in the arg buffer, but rather in a new field and dropping the HasTargetAddress and TargetAddressOffset. In such case, 0 address could indicate an invalid value.
Or is there some reason for storing the address in the buffer?

Reply:
The address is stored in the arg buffer only for `calli` and for generic call cases. In that case it will be passed to `StoreTailCallArgs` as the last argument so that it can be stored there and used by the `CallTailCallTarget` stub. `CallTailCallTarget` will then emit `calli` after loading that address out of the arg buffer. It could maybe be in `TailCallTls` instead, although that would mean we need to somehow get this TLS in the `StoreTailCallArgs` stub. In any case since it is not always used it seems it belongs in the arg buffer.

Reply:
Ah, ok, that was the piece of information I was missing. Makes complete sense then.